sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -348,26 +348,76 @@ private[sql] object DataSourceStrategy extends Strategy with Logging {
         Some(sources.EqualTo(a.name, v))
       case expressions.EqualTo(Literal(v, _), a: Attribute) =>
         Some(sources.EqualTo(a.name, v))
+      case expressions.EqualTo(Cast(a: Attribute, _), Literal(v, _))
+        if a.dataType == DateType || a.dataType == TimestampType =>
+          Some(sources.EqualTo(a.name, v))
```

Comment:
`Cast.eval()` returns values of Catalyst internal types, and can be converted to values of user space types via `CatalystTypeConverter.convertToScala()`. This is not super efficient, but we are not on the critical path, so I think it's OK.